runtime/src/accounts_db.rs

```diff
@@ -2200,33 +2212,156 @@ impl AccountsDb {
+                    break;
+                }
+            }
+            let (new_slot, new_store_id, new_offset) =
+                self.accounts_index_get(ancestors, pubkey, max_root)?;
+            if new_slot == slot && new_store_id == store_id {
+                // If the entry was missing from the cache, that means it must have been flushed,
+                // and the accounts index is always updated before cache flush, so this
+                // should be impossible
+                assert!(new_store_id != CACHE_VIRTUAL_STORAGE_ID);
+
+                // If this is not a cache entry, then this was a minor fork slot
+                // that had its storage entries cleaned up by purge_slots() but hasn't been
+                // cleaned yet, so return None
+                return None;
+            }
+
+            // Accounts cache was just flushed, look in the index to find
+            // the account storage again. This works because in accounts cache flush,
+            // an account is written to storage *before* it is removed from the cache
+            account_accessor = self.get_account_accessor_from_cache_or_storage(
+                new_slot,
+                pubkey,
+                new_store_id,
+                new_offset,
+            );
+        }
+
+        let loaded_account: Option<LoadedAccount> = match &mut account_accessor {
+            LoadedAccountAccessor::Cached(None) | LoadedAccountAccessor::Stored(None) => {
+                panic!("Should have already been taken care of above, see comment above");
+            }
+            LoadedAccountAccessor::Cached(_) => {
+                // Cached(Some(x)) variant always produces `Some` for get_loaded_account() since
+                // it just returns the inner `x` without additional fetches
+                Some(account_accessor.get_loaded_account().unwrap())
+            }
+            LoadedAccountAccessor::Stored(Some(_)) => {
+                let load_result = account_accessor.get_loaded_account();
+                if is_transaction_load {
+                    Some(load_result
+                        .expect("If an account was found in the index, it cannot have been cleaned.
+                        This is because clean should not be removing storage entries that have alive accounts"))
+                } else {
+                    // RPC may have fetched this storage entry after it was already `reset` by clean
+                    load_result
+                }
+            }
         };
 
-        //TODO: thread this as a ref
-        self.get_account_accessor_from_cache_or_storage(slot, pubkey, store_id, offset)
-            .get_loaded_account()
-            .map(|loaded_account| (loaded_account.account(), slot))
+        loaded_account.map(|loaded_account| (loaded_account.account(), slot))
     }
 
     pub fn load_account_hash(&self, ancestors: &Ancestors, pubkey: &Pubkey) -> Hash {
```

Comment:
:red_circle: I thought this bad code duplication might cause bugs. but this is only called from tests.... how about marking this  as `config(test)` or fixing the implementation?

Reply:
Unfortunately it was in `tests/accounts_db` which compiles this `runtime` without test flags, so instead I refactored and fixed the implementation: https://github.com/solana-labs/solana/pull/15454/commits/cdc6f15ec951d8bd5d1e796705fb8a4516a2e1ea